### PR TITLE
create start then load

### PR DESCRIPTION
**Pull request by supertick**

### CODE REVIEW

File: src/main/java/org/myrobotlab/framework/Service.java [hunk 1]
```diff
@@ -2813,5 +2815,15 @@ public boolean isPeerStarted(String peerKey) {
...
     Runtime.getInstance().registerForInterfaceChange(getClass().getCanonicalName(), clazz);
   }
+  
 
+  /**
+   * default emtpy getDefaultConfig - will be overriden by services which provide sane defaults
+   * @return
+   */
+ static public Map<String, ServiceConfig> getDefaultConfig() {
+   Map<String, ServiceConfig> ret = new HashMap<>();
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/framework/Service.java]: i feel like returning null is better than returning an empty config..

File: src/main/java/org/myrobotlab/service/Joystick.java [hunk 2]
```diff
@@ -177,12 +177,14 @@ public void poll() {
         if (pollingController == null) {
           error("controller is null - can not poll");
           stopPolling();
+          break;
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/Joystick.java:180]: does this kill the polling thread?  is this change related to tracking?

File: src/main/java/org/myrobotlab/service/Joystick.java [hunk 3]
```diff
@@ -177,12 +177,14 @@ public void poll() {
...
           error("controller is null - can not poll");
           stopPolling();
+          break;
         }
 
         // get the data
         if (isPolling && !pollingController.poll()) {
           error("failed to poll controller");
           stopPolling();
+          break;
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/Joystick.java:187]: see comment above

File: src/main/java/org/myrobotlab/service/WebGui.java [hunk 4]
```diff
@@ -16,6 +16,8 @@
 import java.util.HashMap;
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/WebGui.java]: this file shouldn't have any changes in this PR.. perhaps we need to merge develop back into this branch?

File: src/main/java/org/myrobotlab/framework/Service.java [hunk 5]
```diff
@@ -2813,5 +2815,15 @@ public boolean isPeerStarted(String peerKey) {
   protected void registerForInterfaceChange(Class<?> clazz) {
     Runtime.getInstance().registerForInterfaceChange(getClass().getCanonicalName(), clazz);
   }
+  
 
+  /**
+   * default emtpy getDefaultConfig - will be overriden by services which provide sane defaults
+   * @return
+   */
+ static public Map<String, ServiceConfig> getDefaultConfig() {
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/framework/Service.java]: I thought we were going to change the return signature here to force a LinkedHashMap ?

File: src/main/java/org/myrobotlab/service/Tracking.java [hunk 6]
```diff
@@ -303,6 +305,66 @@ public ServiceConfig getConfig() {
     return config;
   }
 
+  static public Map<String, ServiceConfig> getDefaultConfig() {
+
+    Map<String, ServiceConfig> config = new HashMap<>();
+
+    RuntimeConfig runtime = new RuntimeConfig();
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/Tracking.java]: i thought we were going to get rid of runtime config here because we are using a linked hash map for the name/service config pairs ..

File: src/main/java/org/myrobotlab/framework/Service.java (hunk 1 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/framework/Service.java]: a bunch of npe safety checks then ... not sure i feel the same way

File: src/main/java/org/myrobotlab/service/Joystick.java (hunk 2 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/service/Joystick.java:180]: 10K ways of implementing a worker ..
it was one of my early implementations ...  it seems kind of goofy ... i should refactor it into my most recent standards ...
but yes the thread will leave run

File: src/main/java/org/myrobotlab/service/WebGui.java (hunk 4 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/service/WebGui.java]: missed import or ctrl-O'd it out in another commit :shrug:

File: src/main/java/org/myrobotlab/framework/Service.java (hunk 5 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/framework/Service.java]: heh .. upcoming commit - we're on the same page

File: src/main/java/org/myrobotlab/service/config/ArduinoConfig.java [hunk 7]
```diff
@@ -1,8 +1,32 @@
...
+
 public class ArduinoConfig extends ServiceConfig {
+  
 
   public String port;
   public boolean connect;
+  public String serial; // name of serial service 
+  
+  @Override
+  public LinkedHashMap<String, ServiceConfig> getDefault(String name) {
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/config/ArduinoConfig.java]: shoudln't this be getDefaultConfig on the Arduino class?   why is it on the ArduinoConfig ?

File: src/main/java/org/myrobotlab/framework/Service.java (hunk 5 above)
Comment (kwatters) [reply src/main/java/org/myrobotlab/framework/Service.java]: This method... should be fixed.. to have the correct return signature and it should be moved to the ServiceInterface.

File: src/main/java/org/myrobotlab/service/config/TrackingConfig.java [hunk 8]
```diff
@@ -9,4 +14,69 @@
   public boolean enabled;
   public long lostTrackingDelayMs = 1000;
 
+  @Override
+  public LinkedHashMap<String, ServiceConfig> getDefault(String name) {
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/config/TrackingConfig.java]: Why is this method here if you have getDefaultConfig on Tracking.java?

File: src/main/java/org/myrobotlab/framework/Service.java [hunk 9]
```diff
@@ -2814,4 +2817,8 @@ protected void registerForInterfaceChange(Class<?> clazz) {
     Runtime.getInstance().registerForInterfaceChange(getClass().getCanonicalName(), clazz);
   }
 
+  public Map<String, ServiceConfig> getDefault() {
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/framework/Service.java]: please correct the return type here. it should be   LinkedHashMap<String, ServiceConfig>   otherwise services could create an unordered configset.

File: src/main/java/org/myrobotlab/service/Runtime.java [hunk 10]
```diff
@@ -3865,4 +3852,104 @@ public void addServiceToRequestedInteface(String serviceName) {
...
+   * @param configPrefixPath
+   *          - prefix location to where the configuration sets will be saved
+   * @param overwrite
+   *          - force overwriting existing config sets
+   * @return - set of successfully saved configuration sets
+   * @throws IOException
+   */
+  static public Set<String> saveDefault(String name, String type, String configPrefixPath, Boolean overwrite) throws IOException {
+
+    Set<String> savedPaths = new HashSet<>();
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/Runtime.java:3900]: Shoudl this be a TreeSet instead of a HashSet?  This is unordered.

File: src/main/java/org/myrobotlab/service/Runtime.java [hunk 11]
```diff
@@ -3865,4 +3852,104 @@ public void addServiceToRequestedInteface(String serviceName) {
...
+        String path = FileIO.gluePaths(dir.getAbsolutePath(), service + ".yml");
+        FileIO.toFile(path, CodecUtils.toYaml(config.get(service)));
+      }
+
+      if (!config.containsKey("runtime")) {
+        // config did not come with an explicit runtime
+        // therefore we will create one with the order of the keyset
+        Map<String, ServiceConfig> runtimeConfig = ServiceInterface.getDefault("runtime", "Runtime");
+        RuntimeConfig rconfig = (RuntimeConfig) runtimeConfig.get("runtime");
+        rconfig.registry = config.keySet().toArray(new String[] {});
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/Runtime.java:3940]: This is unordered here...

File: src/main/java/org/myrobotlab/service/Runtime.java [hunk 12]
```diff
@@ -3865,4 +3852,104 @@ public void addServiceToRequestedInteface(String serviceName) {
...
+
+      for (String service : config.keySet()) {
+        String path = FileIO.gluePaths(dir.getAbsolutePath(), service + ".yml");
+        FileIO.toFile(path, CodecUtils.toYaml(config.get(service)));
+      }
+
+      if (!config.containsKey("runtime")) {
+        // config did not come with an explicit runtime
+        // therefore we will create one with the order of the keyset
+        Map<String, ServiceConfig> runtimeConfig = ServiceInterface.getDefault("runtime", "Runtime");
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/Runtime.java]: to be explicit about the ordering of the keyset here, you probably should make this a LinkedHashMap

File: src/main/java/org/myrobotlab/service/Tracking.java [hunk 13]
```diff
@@ -303,6 +305,68 @@ public ServiceConfig getConfig() {
     return config;
   }
 
+  static public Map<String, ServiceConfig> getDefaultConfig() {
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/Tracking.java:308]: This return type is incorrect,  you should add the @Override annotation here.

Comment (kwatters) [inline src/main/java/org/myrobotlab/service/Tracking.java]: This method should be removed...  you have getDefault() below..   don't put default configs in 2 places on this service.

File: src/main/java/org/myrobotlab/service/Tracking.java [hunk 14]
```diff
@@ -329,64 +393,106 @@ public ServiceConfig load(ServiceConfig c) {
...
+    ServoConfig tiltConfig = (ServoConfig) config.get(trackingConfig.tilt);
+    tiltConfig.pin = "5";
+    tiltConfig.autoDisable = true;
+    tiltConfig.idleTimeout = 3000;
+    tiltConfig.controller = controller;
+
+    PidConfig pidConfig = (PidConfig) config.get(trackingConfig.pid);
+    PidData panData = new PidData();
+    panData.kp = 0.015;
+    panData.ki = 0.005;
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/Tracking.java]: you should be explicit and also define kd here.

File: src/main/java/org/myrobotlab/service/Tracking.java [hunk 15]
```diff
@@ -329,64 +393,106 @@ public ServiceConfig load(ServiceConfig c) {
...
+
+    pidConfig.data.put(trackingConfig.pan, panData);
+
+    PidData tiltData = new PidData();
+    tiltData.kp = 0.015;
+    tiltData.ki = 0.005;
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/Tracking.java]: you should be explicit and also define kd here.

File: src/main/java/org/myrobotlab/service/Tracking.java [hunk 16]
```diff
@@ -329,64 +393,106 @@ public ServiceConfig load(ServiceConfig c) {
...
-      Runtime.start(pid, "Pid");
-      attach(pid);
-    }
-
-    if (cv == null) {
-      cv = String.format("%s.cv", getName());
-      Runtime.start(cv, "OpenCV");
-      attach(cv);
-    }
+  static public LinkedHashMap<String, ServiceConfig> getDefault(String name) {
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/Tracking.java:339]: add the @ Override annotation to this method... you are overriding.

File: src/main/java/org/myrobotlab/service/config/ServiceConfig.java [hunk 17]
```diff
@@ -1,20 +1,33 @@
 package org.myrobotlab.service.config;
 
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Method;
+import java.util.LinkedHashMap;
+import java.util.Map;
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/config/ServiceConfig.java]: cleanup / remove unused imports here.

File: src/main/java/org/myrobotlab/service/Arduino.java [hunk 18]
```diff
@@ -2241,12 +2235,36 @@ public ServiceConfig getConfig() {
...
+      serial = (Serial)Runtime.start(config.serial);
+    }
 
     if (config.port != null) {
       connect(config.port);
     }
     return c;
   }
+  
+  static public LinkedHashMap<String, ServiceConfig> getDefault(String name) {
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/Arduino.java:2250]: Where is the @ Override annotation..  the method signature is wrong.. it should include the type...  as per the interface.

File: src/main/java/org/myrobotlab/service/Tracking.java (hunk 16 above)
Comment (kwatters) [reply src/main/java/org/myrobotlab/service/Tracking.java:339]: Also this method signature isn't correct...  it's missing the Type...   (as per ServiceInterface.)

File: src/main/java/org/myrobotlab/framework/interfaces/ServiceInterface.java [hunk 19]
```diff
@@ -162,4 +172,55 @@
 
   public int getCreationOrder();
 
+  static public LinkedHashMap<String, ServiceConfig> getDefault(String name, String type) {
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/framework/interfaces/ServiceInterface.java:175]: Why do you need to pass Type here?  the service interface contains getType()  ...  you should be able to make that static and use it here also to avoid needing to pass the type in.

Comment (kwatters) [reply src/main/java/org/myrobotlab/framework/interfaces/ServiceInterface.java:175]: i stand semi-corrected...  ignore this.

File: src/main/java/org/myrobotlab/framework/Service.java (hunk 9 above)
Comment (kwatters) [reply src/main/java/org/myrobotlab/framework/Service.java]: this could also be "final" method so that no subclass could override it.

File: src/main/java/org/myrobotlab/framework/interfaces/ServiceInterface.java (hunk 19 above)
Comment (kwatters) [reply src/main/java/org/myrobotlab/framework/interfaces/ServiceInterface.java:175]: actually.. don't ignore this.. you can derive the type of the implementing class with
`public static String getType() {
    return MethodHandles.lookup().lookupClass().getSimpleName();
  };`
